### PR TITLE
[24511] New (X)HTML Syntax for Wiki Macros

### PR DESCRIPTION
**Pull request by silkentrance**
New (X)HTML Syntax for Wiki Macros

THIS IS A WORK IN PROGRESS

see https://community.openproject.com/projects/openproject/work_packages/24511

### CODE REVIEW

**Comment by lindenthal** [comment]
Hi Carsten,

the tests are failing. Can you please first check this?

Thanks
Niels

**Comment by silkentrance** [comment]
@lindenthal Yes, you are right, the tests are still failing. Partly this is because this has been implemented only in part and partly also because I did not care for the helpers ATM. @oliverguenther provided a PR that helped fixing this, but I fear that additional work needs to be done on behalf of these helpers...

**Comment by silkentrance** [comment]
@oliverguenther work was halted due to an orange juice spill on my laptop... I need to figure out a way on how to transfer the local data to the desktop... darn, even rinsing the keyboard with clear warm water did not work...

**Comment by oliverguenther** [comment]
Damn :/ good luck. I'll have another go at the helpers in a quiet minute.

**Comment by silkentrance** [comment]
@oliverguenther Further work on the API and internal workings thereof. MacroState for stateful macros such as IncludeMacro might require more work. Some refactoring done on existing transformers such as RedmineLinkTransformer to make them more easier to refactor and to reduce overall codeclimate issues. Next I will start working on the existing wiki macros such as TocMacro and so on.

**Comment by silkentrance** [comment]
@oliverguenther ~The required refactorings, namely moving from fragment processing to simple text processing and apply fragment processing in only a few use cases increased overall processing speed a lot.~

I moved back to fragment processing as the actual benefit was less than expected.

**Comment by silkentrance** [comment]
@oliverguenther the API is now stable I suppose. The existing macros need some more work, though. I will now begin with fixing the codeclimate issues, then I will fix existing issues with missing helpers and begin reimplementing existing legacy tests and implement required new tests.

The main changes reside in lib/open_project/text_formatting/text_formatting.rb and lib/open_project/text_formatting/* and lib/redmine/wiki_formatting/macros.rb. I also have added view templates for the macro list macro.

Perhaps you can find the time and have a look at the new MacroBase API and the descriptor DSL I added for declaring macro descriptors and also for registering the macro with the system. See for example lib/open_project/text_formatting/macros/provided/toc.rb. (Note: the sources in there will be renamed to e.g. toc_macro.rb to comply with existing auto loading rules.)

File: lib/open_project/text_formatting.rb [hunk 1]
```diff
@@ -56,473 +44,122 @@ def truncate_lines(string, options = {})
...
+          OpenProject::TextFormatting::Internal::Transformers::RedmineWikiTransformer.new(view),
+          OpenProject::TextFormatting::Internal::Transformers::RedmineLinkTransformer.new(view),
+          OpenProject::TextFormatting::Internal::Transformers::WikiLinkTransformer.new(view),
+          # must run after redmine*/wiki link transformers during the
+          # process stage
+          OpenProject::TextFormatting::Internal::Transformers::MacroTransformer.new(view),
+          # the below two must run after macro transformer during the
+          # post process stage
+          OpenProject::TextFormatting::Internal::Transformers::ImageAttachmentTransformer.new(view),
+          # WTF?: all occurrences of {{ have already been replaced by {{ $root... }} by whatever
```
Comment (oliverguenther) [inline lib/open_project/text_formatting.rb:96]: This is caused by `angular-xss` gem which patches ERB and CGI.escape to do this transformation from non `html_safe` output to receive this transformation.

**Comment by oliverguenther** [comment]
Hi @silkentrance, on the last commit, macros aren't expanded in their existing `{{..}}` notation. I don't mind the usage of a DSL, but the amount of instance mutations on the `macro_base` are hard to read and I fear, hard to maintain. I would prefer to reduce the complexity of the registry by required regular classes simply registering themselves with the registry as an initializer. A base class could still be extracted and implemented by macros.

Please do pitch me the idea of a DSL if you truly believe it is superior, but currently I'm not sold on that ;-)

**Comment by silkentrance** [comment]
@oliverguenther Ups, the legacy notation should be supported. I will have a look at this tomorrow.

As for the DSL, well it is a first try. I also tried using YAML files, which would be of the same name as the ruby module and located in the same folder. This could then be loaded on register or on self.inherited, providing a getter only for the descriptor.

What do you mean by ```simply registering themselves with the registry as an initializer```?

**Comment by oliverguenther** [comment]
Closing this for now since neither Carsten nor I am actively working on it.

**Pull request closed without merging**